File: python/sony_metadata.py
# sony_metadata.py
"""Extractie profunda de metadate Sony (Log/Gamma, EI, ISO/expunere/diafragma
per-cadru, XML sidecar) — port 1:1 al `SonyMetadata.swift` (Mac), care la
randul lui e portat din parserul JS scris de mana in GDC_Metadata_View_Premium.
Vezi SonyMetadata.swift pentru comentariile complete de design (pista rtmd e
un format binar Sony nedocumentat oficial, dedus prin efort al comunitatii)."""

import os
import struct
import xml.etree.ElementTree as ET

CONTAINER_TYPES = {"moov", "trak", "mdia", "minf", "stbl", "udta", "edts", "dinf", "meta"}

EXPOSURE_MODE_LABELS = {
    0x01010000: "Manual", 0x01020000: "Auto complet",
    0x01030000: "Auto cu prioritate câștig", 0x01040000: "Auto cu prioritate diafragmă",
    0x01050000: "Auto cu prioritate timp expunere",
}

ITEM_LABELS = {
    "CaptureGammaEquation": "Curbă Gamma (Log)", "CaptureColorPrimaries": "Gamut culoare (Log)",
    "CodingEquations": "Ecuații de codare (matrice)", "CaptureFrameRate": "Cadre/s la captură",
    "CaptureBitDepth": "Adâncime biți captură", "CodingEIFlag": "Flag EI (Exposure Index)",
    "ExposureIndexOfPictureProfile": "Exposure Index (EI)", "WhiteBalance": "Balans de alb",
    "ColorTemperature": "Temperatură culoare", "ElectricalExtenderMagnification": "Extender electronic",
    "ImagerDimension": "Dimensiune senzor", "MasterBlackLevel": "Nivel negru master",
    "MasterGainAdjustment": "Ajustare câștig master", "ImagerScanMode": "Mod scanare senzor",
    "AutoSlowShutter": "Slow shutter automat", "NDFilter": "Filtru ND",
}


def read(path):
    """Intoarce {"camera_profile": {...}, "capture_settings": {...}} — gol
    daca fisierul nu e Sony / nu are nimic de extras (esec silentios, ca in
    JS — asta nu e un fisier obligatoriu de citit cu succes)."""
    ext = os.path.splitext(path)[1].lower().lstrip(".")
    result = {"camera_profile": {}, "capture_settings": {}}

    if ext == "xml":
        try:
            with open(path, "r", encoding="utf-8", errors="ignore") as f:
                _parse_sony_xml(f.read(), result["camera_profile"])
        except OSError:
            pass
        return result

    if ext not in ("mp4", "mov", "m4v"):
        return result

    try:
        with open(path, "rb") as f:
            f.seek(0, os.SEEK_END)
            file_size = f.tell()
            f.seek(0)

            state = {"xml_text": None, "handlers": [], "rtmd_offset": None, "rtmd_size": None}

            def on_box(box):
                if box["type"] == "hdlr":
                    handler = _read(f, box["content_start"] + 8, 4)
                    if handler:
                        state["handlers"].append(handler.decode("ascii", "ignore"))
                last_handler = state["handlers"][-1] if state["handlers"] else None

                if last_handler == "rtmd" and state["rtmd_size"] is None and box["type"] == "stsz":
                    head = _read(f, box["content_start"] + 4, 8)
                    if head:
                        sample_size, sample_count = struct.unpack(">II", head)
                        if sample_size != 0:
                            state["rtmd_size"] = sample_size
                        elif sample_count > 0:
                            first = _read(f, box["content_start"] + 12, 4)
                            if first:
                                state["rtmd_size"] = struct.unpack(">I", first)[0]

                if last_handler == "rtmd" and state["rtmd_offset"] is None and box["type"] in ("stco", "co64"):
                    count_bytes = _read(f, box["content_start"] + 4, 4)
                    if count_bytes and struct.unpack(">I", count_bytes)[0] > 0:
                        if box["type"] == "stco":
                            off_bytes = _read(f, box["content_start"] + 8, 4)
                            if off_bytes:
                                state["rtmd_offset"] = struct.unpack(">I", off_bytes)[0]
                        else:
                            off_bytes = _read(f, box["content_start"] + 8, 8)
                            if off_bytes:
                                state["rtmd_offset"] = struct.unpack(">Q", off_bytes)[0]

                if box["type"] == "meta" and state["xml_text"] is None:
                    state["xml_text"] = _try_extract_sony_xml_box(f, box)

            _walk_iso_boxes(f, 0, file_size, on_box)

            if state["xml_text"]:
                _parse_sony_xml(state["xml_text"], result["camera_profile"])
            if state["rtmd_offset"] is not None and state["rtmd_size"]:
                sample = _read(f, state["rtmd_offset"], state["rtmd_size"])
                if sample:
                    result["capture_settings"] = _parse_sony_rtmd_sample(sample)
    except OSError:
        pass

    return result


# ---------- cutii ISO-BMFF ----------

def _read(f, offset, count):
    if count <= 0:
        return None
    f.seek(offset)
    data = f.read(count)
    return data if len(data) == count else None


def _read_box_header(f, offset):
    head = _read(f, offset, 8)
    if not head:
        return None
    size = struct.unpack(">I", head[0:4])[0]
    box_type = head[4:8].decode("ascii", "ignore")
    header_size = 8
    if size == 1:
        ext = _read(f, offset + 8, 8)
        if not ext:
            return None
        size = struct.unpack(">Q", ext)[0]
        header_size = 16
    elif size == 0:
        return None
    if size < 8:
        return None
    return {"type": box_type, "size": size, "start": offset,
            "content_start": offset + header_size, "end": offset + size}


def _walk_iso_boxes(f, start, end, on_box):
    offset = start
    while offset + 8 <= end:
        box = _read_box_header(f, offset)
        if not box or box["end"] > end or box["end"] <= box["start"]:
            break
        on_box(box)
        if box["type"] in CONTAINER_TYPES:
            child_start = box["content_start"]
            if box["type"] == "meta":
                if _try_extract_sony_xml_box(f, box) is None:
                    child_start += 4
                else:
                    child_start = None
            if child_start is not None:
                _walk_iso_boxes(f, child_start, box["end"], on_box)
        offset = box["end"]


def _try_extract_sony_xml_box(f, box):
    length = box["end"] - box["content_start"]
    if length <= 0 or length >= 8 * 1024 * 1024:
        return None
    raw = _read(f, box["content_start"], length)
    if not raw:
        return None
    try:
        text = raw.decode("utf-8", errors="ignore")
    except UnicodeDecodeError:
        return None
    idx = text.find("<?xml")
    if idx == -1:
        return None
    nul_idx = text.find("\x00", idx)
    xml_part = text[idx:] if nul_idx == -1 else text[idx:nul_idx]
    return xml_part.strip()


# ---------- decodor rtmd (KLV binar, primul esantion) ----------

def _parse_sony_rtmd_sample(sample):
    pos = 0
    end = len(sample)
    if end >= 17 and sample[0] == 0x06:
        pos = 16
        b0 = sample[pos]
        pos += (1 + (b0 & 0x7F)) if (b0 & 0x80) else 1
    out = {}
    while pos + 4 <= end:
        tag, length = struct.unpack(">HH", sample[pos:pos + 4])
        pos += 4
        if pos + length > end:
            break
        label_value = _decode_rtmd_field(tag, sample, pos, length)
        if label_value and label_value[0] not in out:
            out[label_value[0]] = label_value[1]
        pos += length
    return out


def _decode_rtmd_field(tag, buf, offset, length):
    chunk = buf[offset:offset + length]
    try:
        if tag == 0x810B and length >= 2:
            return ("ISO (rtmd, primul cadru)", str(struct.unpack(">H", chunk[:2])[0]))
        if tag in (0x8119, 0xE301) and length >= 4:
            return ("ISO (rtmd, primul cadru)", str(struct.unpack(">I", chunk[:4])[0]))
        if tag == 0x8109 and length >= 8:
            num, den = struct.unpack(">II", chunk[:8])
            if num and den:
                return ("Timp expunere (rtmd, primul cadru)", f"1/{round(den / num)} s")
            return None
        if tag == 0x8000 and length >= 2:
            raw = struct.unpack(">H", chunk[:2])[0]
            fstop = 2 ** (8 * (1 - raw / 65536))
            return ("Diafragmă (rtmd, primul cadru)", f"f/{fstop:.1f}")
        if tag == 0x810E and length >= 2:
            return ("Balans de alb (rtmd, primul cadru)", f"{struct.unpack('>H', chunk[:2])[0]} K")
        if tag == 0x8100 and length >= 16:
            mode = struct.unpack(">I", chunk[12:16])[0]
            label = EXPOSURE_MODE_LABELS.get(mode)
            return ("Mod expunere (rtmd, primul cadru)", label) if label else None
        if tag == 0x8106 and length >= 8:
            num, den = struct.unpack(">II", chunk[:8])
            if den:
                return ("Cadre/s captură (rtmd, primul cadru)", f"{num / den:.2f}")
            return None
    except struct.error:
        return None
    return None


# ---------- XML Sony (sidecar sau embedat) ----------

def _attr(el, *names):
    for n in names:
        v = el.get(n)
        if v:
            return v
    return None


def _parse_sony_xml(text, cat):
    try:
        root = ET.fromstring(text)
    except ET.ParseError:
        return

    device = root.find(".//Device")
    if device is not None:
        man = _attr(device, "manufacturer", "Manufacturer")
        model = _attr(device, "modelName", "ModelName")
        serial = _attr(device, "serialNo", "SerialNo")
        if man:
            cat["Producător cameră"] = man
        if model:
            cat["Model cameră"] = model
        if serial:
            cat["Serie cameră"] = serial

    creation_date = root.find(".//CreationDate")
    if creation_date is not None:
        v = _attr(creation_date, "value", "Value")
        if v:
            cat["Data creare clip"] = v

    video_frame = root.find(".//VideoFrame")
    if video_frame is not None:
        codec = _attr(video_frame, "videoCodec", "VideoCodec")
        cap_fps = _attr(video_frame, "captureFps", "CaptureFps")
        fmt_fps = _attr(video_frame, "formatFps", "FormatFps")
        aspect = _attr(video_frame, "aspectRatio", "AspectRatio")
        if codec:
            cat["Codec video (XML)"] = codec
        if cap_fps:
            cat["FPS captură (XML)"] = cap_fps
        if fmt_fps:
            cat["FPS format (XML)"] = fmt_fps
        if aspect:
            cat["Aspect ratio (XML)"] = aspect

    video_layout = root.find(".//VideoLayout")
    if video_layout is not None:
        w = _attr(video_layout, "pixel", "Pixel")
        h = _attr(video_layout, "numOfVerticalLine", "NumOfVerticalLine")
        if w and h:
            cat["Rezoluție (XML)"] = f"{w} x {h}"

    for item in root.iter("Item"):
        name = _attr(item, "name", "Name")
        value = _attr(item, "value", "Value")
        if not name or not value:
            continue
        label = ITEM_LABELS.get(name, name)
        if label not in cat:
            cat[label] = value

File: CGConvertor/SonyMetadata.swift
// SonyMetadata.swift
// Extractie profunda de metadate specifice camerelor Sony — profil Log/Gamma,
// EI, si o aproximare a ISO/expunere/diafragma/balans de alb la momentul
// capturii, citite direct din containerul ISO-BMFF (MP4/MOV) sau dintr-un XML
// sidecar Sony separat. Portat 1:1 (logica, nu sintaxa) din parserul JS scris
// de mana in GDC_Metadata_View_Premium (~/Developer/GDC_Metadata_View_Premium/
// index.html) — acolo functiona pe File/ArrayBuffer in browser; aici citim
// direct de pe disc cu FileHandle, ceea ce e de fapt mai simplu (nu mai
// trebuie sa navigam sloturi de File.slice()).
//
// Pista "rtmd" (metadate per-cadru Sony) e un format binar KLV nedocumentat
// oficial de Sony, dedus prin efortul comunitatii open-source (vezi proiectul
// telemetry-parser) — citim DOAR primul esantion (primul cadru), suficient
// pentru o aproximare a setarilor de captura, nu o extractie completa
// cadru-cu-cadru (ar fi mult mai scump si nu aduce valoare pentru un
// instrument de inspectie/comparatie, nu de analiza per-cadru).

import Foundation

struct SonyMetadata {
    /// Chei -> valori, gata de afisat (eticheta RO deja aplicata, ca la restul aplicatiei).
    var cameraProfile: [String: String] = [:]
    var captureSettings: [String: String] = [:]

    var isEmpty: Bool { cameraProfile.isEmpty && captureSettings.isEmpty }
}

enum SonyMetadataReader {

    // MARK: - Punct de intrare

    /// `url` poate fi un clip video (se cauta XML embedat + pista rtmd) sau
    /// direct un fisier `.xml` sidecar Sony.
    static func read(from url: URL) -> SonyMetadata {
        var result = SonyMetadata()
        let ext = url.pathExtension.lowercased()

        if ext == "xml" {
            if let text = try? String(contentsOf: url, encoding: .utf8) {
                parseSonyXML(text, into: &result.cameraProfile)
            }
            return result
        }

        guard ["mp4", "mov", "m4v"].contains(ext),
              let handle = try? FileHandle(forReadingFrom: url) else { return result }
        defer { try? handle.close() }

        guard let fileSize = try? handle.seekToEnd() else { return result }
        handle.seek(toFileOffset: 0)

        var xmlText: String?
        var trackHandlers: [String] = []
        var rtmdSampleOffset: UInt64?
        var rtmdSampleSize: UInt32?

        walkIsoBoxes(handle: handle, start: 0, end: fileSize) { box in
            if box.type == "hdlr", let bytes = readBytes(handle, at: box.contentStart + 8, count: 4),
               let handlerType = String(bytes: bytes, encoding: .ascii) {
                trackHandlers.append(handlerType)
            }
            let lastHandler = trackHandlers.last

            if lastHandler == "rtmd", rtmdSampleSize == nil, box.type == "stsz" {
                if let bytes = readBytes(handle, at: box.contentStart + 4, count: 8) {
                    let sampleSize = readUInt32BE(bytes, 0)
                    let sampleCount = readUInt32BE(bytes, 4)
                    if sampleSize != 0 {
                        rtmdSampleSize = sampleSize
                    } else if sampleCount > 0, let sizeBytes = readBytes(handle, at: box.contentStart + 12, count: 4) {
                        rtmdSampleSize = readUInt32BE(sizeBytes, 0)
                    }
                }
            }
            if lastHandler == "rtmd", rtmdSampleOffset == nil, box.type == "stco" || box.type == "co64" {
                if let countBytes = readBytes(handle, at: box.contentStart + 4, count: 4) {
                    let entryCount = readUInt32BE(countBytes, 0)
                    if entryCount > 0 {
                        if box.type == "stco", let offBytes = readBytes(handle, at: box.contentStart + 8, count: 4) {
                            rtmdSampleOffset = UInt64(readUInt32BE(offBytes, 0))
                        } else if let offBytes = readBytes(handle, at: box.contentStart + 8, count: 8) {
                            rtmdSampleOffset = readUInt64BE(offBytes, 0)
                        }
                    }
                }
            }
            if box.type == "meta", xmlText == nil {
                xmlText = tryExtractSonyXMLBox(handle, box)
            }
        }

        if let text = xmlText {
            parseSonyXML(text, into: &result.cameraProfile)
        }
        if let offset = rtmdSampleOffset, let size = rtmdSampleSize, size > 0,
           let sampleBytes = readBytes(handle, at: offset, count: Int(size)) {
            result.captureSettings = parseSonyRtmdSample(sampleBytes)
        }

        return result
    }

    // MARK: - Cutii ISO-BMFF (parser generic de container MP4/MOV)

    private struct IsoBox {
        let type: String
        let size: UInt64
        let start: UInt64
        let contentStart: UInt64
        let end: UInt64
    }

    private static let containerTypes: Set<String> =
        ["moov", "trak", "mdia", "minf", "stbl", "udta", "edts", "dinf", "meta"]

    private static func readBytes(_ handle: FileHandle, at offset: UInt64, count: Int) -> [UInt8]? {
        guard count > 0 else { return nil }
        handle.seek(toFileOffset: offset)
        guard let data = try? handle.read(upToCount: count), data.count == count else { return nil }
        return [UInt8](data)
    }

    private static func readUInt32BE(_ bytes: [UInt8], _ offset: Int) -> UInt32 {
        (UInt32(bytes[offset]) << 24) | (UInt32(bytes[offset + 1]) << 16) |
        (UInt32(bytes[offset + 2]) << 8) | UInt32(bytes[offset + 3])
    }

    private static func readUInt16BE(_ bytes: [UInt8], _ offset: Int) -> UInt16 {
        (UInt16(bytes[offset]) << 8) | UInt16(bytes[offset + 1])
    }

    private static func readUInt64BE(_ bytes: [UInt8], _ offset: Int) -> UInt64 {
        var v: UInt64 = 0
        for i in 0..<8 { v = (v << 8) | UInt64(bytes[offset + i]) }
        return v
    }

    private static func readBoxHeader(_ handle: FileHandle, at offset: UInt64) -> IsoBox? {
        guard let head = readBytes(handle, at: offset, count: 8) else { return nil }
        var size = UInt64(readUInt32BE(head, 0))
        guard let type = String(bytes: head[4..<8], encoding: .ascii) else { return nil }
        var headerSize: UInt64 = 8
        if size == 1 {
            guard let ext = readBytes(handle, at: offset + 8, count: 8) else { return nil }
            size = readUInt64BE(ext, 0)
            headerSize = 16
        } else if size == 0 {
            return nil // "pana la sfarsitul fisierului" — nu se intampla la boxurile pe care le cautam
        }
        guard size >= 8 else { return nil }
        return IsoBox(type: type, size: size, start: offset,
                      contentStart: offset + headerSize, end: offset + size)
    }

    /// Traverseaza recursiv arborele de cutii, apeland `body` pentru fiecare —
    /// identic ca strategie cu `walkIsoBoxes` din JS.
    private static func walkIsoBoxes(handle: FileHandle, start: UInt64, end: UInt64, body: (IsoBox) -> Void) {
        var offset = start
        while offset + 8 <= end {
            guard let box = readBoxHeader(handle, at: offset), box.end <= end, box.end > box.start else { break }
            body(box)
            if containerTypes.contains(box.type) {
                var childStart: UInt64? = box.contentStart
                if box.type == "meta" {
                    // boxul "meta" clasic (non-QuickTime) are 4 octeti de flags
                    // inainte de copii — dar daca gasim XML Sony direct in
                    // continutul lui, tratam boxul ca frunza (nu mai coboram).
                    if tryExtractSonyXMLBox(handle, box) == nil {
                        childStart = box.contentStart + 4
                    } else {
                        childStart = nil
                    }
                }
                if let childStart {
                    walkIsoBoxes(handle: handle, start: childStart, end: box.end, body: body)
                }
            }
            offset = box.end
        }
    }

    private static func tryExtractSonyXMLBox(_ handle: FileHandle, _ box: IsoBox) -> String? {
        let length = Int(box.end - box.contentStart)
        guard length > 0, length < 8 * 1024 * 1024, let bytes = readBytes(handle, at: box.contentStart, count: length) else { return nil }
        guard let text = String(bytes: bytes, encoding: .utf8) else { return nil }
        guard let range = text.range(of: "<?xml") else { return nil }
        var xmlPart = String(text[range.lowerBound...])
        if let nul = xmlPart.firstIndex(of: "\0") { xmlPart = String(xmlPart[..<nul]) }
        return xmlPart.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Decodor rtmd (KLV binar, primul esantion/cadru)

    private static let exposureModeLabels: [UInt32: String] = [
        0x01010000: "Manual", 0x01020000: "Auto complet",
        0x01030000: "Auto cu prioritate câștig", 0x01040000: "Auto cu prioritate diafragmă",
        0x01050000: "Auto cu prioritate timp expunere",
    ]

    private static func parseSonyRtmdSample(_ bytes: [UInt8]) -> [String: String] {
        var pos = 0
        let end = bytes.count
        // primul bloc rtmd contine un header cu marker 0x06 urmat de un
        // "context data set" de lungime variabila (BER) — il sarim, nu ne
        // intereseaza continutul lui, doar KLV-urile Sony de dupa.
        if end >= 17, bytes[0] == 0x06 {
            pos = 16
            let b0 = bytes[pos]
            pos += (b0 & 0x80) != 0 ? Int(1 + (b0 & 0x7f)) : 1
        }
        var out: [String: String] = [:]
        while pos + 4 <= end {
            let tag = readUInt16BE(bytes, pos)
            let len = Int(readUInt16BE(bytes, pos + 2))
            pos += 4
            guard pos + len <= end else { break }
            if let (label, value) = decodeRtmdField(tag: tag, bytes: bytes, offset: pos, length: len),
               out[label] == nil {
                out[label] = value
            }
            pos += len
        }
        return out
    }

    private static func decodeRtmdField(tag: UInt16, bytes: [UInt8], offset: Int, length: Int) -> (String, String)? {
        switch tag {
        case 0x810B where length >= 2:
            return ("ISO (rtmd, primul cadru)", String(readUInt16BE(bytes, offset)))
        case 0x8119 where length >= 4, 0xE301 where length >= 4:
            return ("ISO (rtmd, primul cadru)", String(readUInt32BE(bytes, offset)))
        case 0x8109 where length >= 8:
            let num = readUInt32BE(bytes, offset), den = readUInt32BE(bytes, offset + 4)
            guard num != 0, den != 0 else { return nil }
            return ("Timp expunere (rtmd, primul cadru)", "1/\(Int((Double(den) / Double(num)).rounded())) s")
        case 0x8000 where length >= 2:
            let raw = readUInt16BE(bytes, offset)
            let fstop = pow(2.0, 8.0 * (1.0 - Double(raw) / 65536.0))
            return ("Diafragmă (rtmd, primul cadru)", String(format: "f/%.1f", fstop))
        case 0x810E where length >= 2:
            return ("Balans de alb (rtmd, primul cadru)", "\(readUInt16BE(bytes, offset)) K")
        case 0x8100 where length >= 16:
            let mode = readUInt32BE(bytes, offset + 12)
            guard let label = exposureModeLabels[mode] else { return nil }
            return ("Mod expunere (rtmd, primul cadru)", label)
        case 0x8106 where length >= 8:
            let num = readUInt32BE(bytes, offset), den = readUInt32BE(bytes, offset + 4)
            guard den != 0 else { return nil }
            return ("Cadre/s captură (rtmd, primul cadru)", String(format: "%.2f", Double(num) / Double(den)))
        default:
            return nil
        }
    }

    // MARK: - XML Sony (sidecar sau embedat) — parsare tolerantă cu XMLParser

    private static let itemLabels: [String: String] = [
        "CaptureGammaEquation": "Curbă Gamma (Log)", "CaptureColorPrimaries": "Gamut culoare (Log)",
        "CodingEquations": "Ecuații de codare (matrice)", "CaptureFrameRate": "Cadre/s la captură",
        "CaptureBitDepth": "Adâncime biți captură", "CodingEIFlag": "Flag EI (Exposure Index)",
        "ExposureIndexOfPictureProfile": "Exposure Index (EI)", "WhiteBalance": "Balans de alb",
        "ColorTemperature": "Temperatură culoare", "ElectricalExtenderMagnification": "Extender electronic",
        "ImagerDimension": "Dimensiune senzor", "MasterBlackLevel": "Nivel negru master",
        "MasterGainAdjustment": "Ajustare câștig master", "ImagerScanMode": "Mod scanare senzor",
        "AutoSlowShutter": "Slow shutter automat", "NDFilter": "Filtru ND",
    ]

    private static func parseSonyXML(_ text: String, into cat: inout [String: String]) {
        guard let data = text.data(using: .utf8) else { return }
        let delegate = SonyXMLParserDelegate()
        let parser = XMLParser(data: data)
        parser.delegate = delegate
        guard parser.parse() else { return }

        if let man = delegate.deviceAttrs["manufacturer"] ?? delegate.deviceAttrs["Manufacturer"] {
            cat["Producător cameră"] = man
        }
        if let model = delegate.deviceAttrs["modelName"] ?? delegate.deviceAttrs["ModelName"] {
            cat["Model cameră"] = model
        }
        if let serial = delegate.deviceAttrs["serialNo"] ?? delegate.deviceAttrs["SerialNo"] {
            cat["Serie cameră"] = serial
        }
        if let creationDate = delegate.creationDate { cat["Data creare clip"] = creationDate }
        if let codec = delegate.videoFrameAttrs["videoCodec"] ?? delegate.videoFrameAttrs["VideoCodec"] {
            cat["Codec video (XML)"] = codec
        }
        if let fps = delegate.videoFrameAttrs["captureFps"] ?? delegate.videoFrameAttrs["CaptureFps"] {
            cat["FPS captură (XML)"] = fps
        }
        if let fps = delegate.videoFrameAttrs["formatFps"] ?? delegate.videoFrameAttrs["FormatFps"] {
            cat["FPS format (XML)"] = fps
        }
        if let aspect = delegate.videoFrameAttrs["aspectRatio"] ?? delegate.videoFrameAttrs["AspectRatio"] {
            cat["Aspect ratio (XML)"] = aspect
        }
        if let w = delegate.videoLayoutAttrs["pixel"] ?? delegate.videoLayoutAttrs["Pixel"],
           let h = delegate.videoLayoutAttrs["numOfVerticalLine"] ?? delegate.videoLayoutAttrs["NumOfVerticalLine"] {
            cat["Rezoluție (XML)"] = "\(w) x \(h)"
        }
        for (name, value) in delegate.items where cat[itemLabels[name] ?? name] == nil {
            cat[itemLabels[name] ?? name] = value
        }
    }

    private final class SonyXMLParserDelegate: NSObject, XMLParserDelegate {
        var deviceAttrs: [String: String] = [:]
        var creationDate: String?
        var videoFrameAttrs: [String: String] = [:]
        var videoLayoutAttrs: [String: String] = [:]
        var items: [(String, String)] = []

        func parser(_ parser: XMLParser, didStartElement elementName: String, namespaceURI: String?,
                    qualifiedName qName: String?, attributes attributeDict: [String: String]) {
            switch elementName {
            case "Device": deviceAttrs = attributeDict
            case "CreationDate": creationDate = attributeDict["value"] ?? attributeDict["Value"]
            case "VideoFrame": videoFrameAttrs = attributeDict
            case "VideoLayout": videoLayoutAttrs = attributeDict
            case "Item":
                if let name = attributeDict["name"] ?? attributeDict["Name"],
                   let value = attributeDict["value"] ?? attributeDict["Value"] {
                    items.append((name, value))
                }
            default: break
            }
        }
    }
}
